runtime/extern.c

```diff
@@ -536,15 +536,35 @@ static void extern_rec(value v)
       continue;
     case Custom_tag: {
       uintnat sz_32, sz_64;
+      char * size_header;
       char * ident = Custom_ops_val(v)->identifier;
       void (*serialize)(value v, uintnat * bsize_32,
                         uintnat * bsize_64)
         = Custom_ops_val(v)->serialize;
+      const struct custom_fixed_length* fixed_length
+        = Custom_ops_val(v)->fixed_length;
       if (serialize == NULL)
         extern_invalid_argument("output_value: abstract value (Custom)");
-      write(CODE_CUSTOM);
-      writeblock(ident, strlen(ident) + 1);
-      Custom_ops_val(v)->serialize(v, &sz_32, &sz_64);
+      if (!fixed_length) {
+        write(CODE_CUSTOM_LEN);
+        writeblock(ident, strlen(ident) + 1);
+        /* Reserve 12 bytes for the lengths (sz_32 and sz_64). */
+        if (extern_ptr + 12 >= extern_limit) grow_extern_output(12);
+        size_header = extern_ptr;
+        extern_ptr += 12;
+        Custom_ops_val(v)->serialize(v, &sz_32, &sz_64);
+        /* Store length before serialized block */
```

Comment:
You could use the `serialize` variable rather than fetch from `Custom_ops_val(v)` again.  CSE in the C compiler might not be clever enough to reuse it for you.

Reply:
Fixed